### ISSUE TITLE
Forked child processes created through rails runner throw error from ActiveRecord when exiting

### ISSUE THREAD

**Issue by gpallotta** [rails/rails#51298]
In Rails 7.1.3.2, child processes which were created using a fork call throw the following error when they are exiting:

```
.../activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:197:in `active_connection?': undefined method `[]' for nil (NoMethodError)

        @thread_cached_conns[connection_cache_key(current_thread)]
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

These child processes are created using a fork call which does not take a block. Notably, if a block is passed to the fork call then the error does not occur.

### Steps to reproduce
```
> ruby -v
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [x86_64-linux]

> rails new fork_child_process -G -M -C -A -J -T 
rails (7.1.3.2)

> echo "fork" > script.rb; bundle exec rails runner script.rb
/home/gpallotta/.gem/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:197:in `active_connection?': undefined method `[]' for nil (NoMethodError)

        @thread_cached_conns[connection_cache_key(current_thread)]
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        from /home/gpallotta/.gem/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/query_cache.rb:45:in `disable_query_cache!'
	from /home/gpallotta/.gem/gems/activerecord-7.1.3.2/lib/active_record/query_cache.rb:33:in `block in complete'
	from /home/gpallotta/.gem/gems/activerecord-7.1.3.2/lib/active_record/query_cache.rb:33:in `each'
	from /home/gpallotta/.gem/gems/activerecord-7.1.3.2/lib/active_record/query_cache.rb:33:in `complete'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/execution_wrapper.rb:37:in `before'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:426:in `block in make_lambda'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:202:in `block (2 levels) in halting'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:707:in `block (2 levels) in default_terminator'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:706:in `catch'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:706:in `block in default_terminator'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:203:in `block in halting'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:598:in `block in invoke_before'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:598:in `each'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:598:in `invoke_before'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:109:in `run_callbacks'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/execution_wrapper.rb:143:in `complete'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/execution_wrapper.rb:137:in `complete!'
	from /home/gpallotta/.gem/gems/activesupport-7.1.3.2/lib/active_support/execution_wrapper.rb:97:in `wrap'
	from /home/gpallotta/.gem/gems/railties-7.1.3.2/lib/rails/commands/runner/runner_command.rb:40:in `perform'
	from /home/gpallotta/.gem/gems/thor-1.3.1/lib/thor/command.rb:28:in `run'
	from /home/gpallotta/.gem/gems/thor-1.3.1/lib/thor/invocation.rb:127:in `invoke_command'
	from /home/gpallotta/.gem/gems/railties-7.1.3.2/lib/rails/command/base.rb:178:in `invoke_command'
	from /home/gpallotta/.gem/gems/thor-1.3.1/lib/thor.rb:527:in `dispatch'
	from /home/gpallotta/.gem/gems/railties-7.1.3.2/lib/rails/command/base.rb:73:in `perform'
	from /home/gpallotta/.gem/gems/railties-7.1.3.2/lib/rails/command.rb:71:in `block in invoke'
	from /home/gpallotta/.gem/gems/railties-7.1.3.2/lib/rails/command.rb:149:in `with_argv'
	from /home/gpallotta/.gem/gems/railties-7.1.3.2/lib/rails/command.rb:69:in `invoke'
	from /home/gpallotta/.gem/gems/railties-7.1.3.2/lib/rails/commands.rb:18:in `<main>'
	from /opt/ruby-3.3.0-1/lib/ruby/3.3.0/bundled_gems.rb:74:in `require'
	from /opt/ruby-3.3.0-1/lib/ruby/3.3.0/bundled_gems.rb:74:in `block (2 levels) in replace_require'
	from /home/gpallotta/.gem/gems/bootsnap-1.18.3/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	from bin/rails:4:in `<main>'

>
```

If instead a block is given to the fork call, then no error occurs

```
> echo "fork {}" > script.rb; bundle exec rails runner script.rb

>
```

### Expected behavior
The forked child process exits cleanly.

### Actual behavior
The forked child process exits with an error.

### System configuration
**Rails version**: 7.1.3.2
**Ruby version**: 3.3.0

**Comment by JoeDupuis** [comment]
Still trying to build a one file repro script, but this is affected only `~> 7.1`. 
`~> 7.0` and `~> 7.2` are not affected.

It looks like the first process to returns kills the db connection and the other process fails. This would explain why passing a block doesn't crash. This script also doesn't crash:

```ruby
pid = fork
if pid.nil?
  sleep 1
else
  Process.kill("HUP",pid)
end
```
Remove the `Process.kill` and it does crash.

**Comment by JoeDupuis** [comment]
I was wrong, I updated my last comment, 7.0 is affected too.

**Comment by JoeDupuis** [comment]
Here's a one file repro:
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", "7.1.3.2"
  gem "sqlite3"
  # If you want to test against edge Rails replace the previous line with this:
  # gem "rails", github: "rails/rails", branch: "main"
end

require "action_controller/railtie"
require "rails/command"
require "active_record"

class TestApp < Rails::Application
  config.root = __dir__
  config.hosts << "example.org"
  config.secret_key_base = "secret_key_base"
  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger
end


ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)


require "minitest/autorun"
require "rack/test"

class BugTest < Minitest::Test
  def test_returns_success

    ActiveRecord::QueryCache.install_executor_hooks
    Rails.application.executor.wrap(source: "application.runner.railties") do
      fork
    end
  end

  private
    def app
      Rails.application
    end
end
```

I have a PR coming for a potential fix.

**Comment by JoeDupuis** [comment]
PR for the fix #51298
